### ISSUE TITLE
wrap multiple statements before arrow in a `do` block

### ISSUE THREAD

**Issue by hoebat** [jimpil/fudje#6]
sometimes I do:

```
(fact 
   (setup obj)
   (call obj)
=> <check>)
```

and this is currently giving a error because the checks expect `<input>`, `=>` and `<output>`. It would be great if multiple calls are supported.

**Comment by hoebat** [comment]
as well as isolated statements in the `fact` form - ie:

```
(fact 
   (comment))
```

**Comment by jimpil** [comment]
Before giving you the full explanation why this can't be done, do you mind me asking, why does the `(setup obj)` expression need to be underneath the `fact` rather than above it? In my experience, the only reason for wanting to do that, is to use meta-constants within the `(setup obj)` expression. For example, you might want to have something like this:

``` clj
(fact "testing whatever" 
  (setup ...a... ...b... ...c...)
  (f1 ...a...) => :some-value
  (f2 ...b...) => :some-other-value  
)
```

The way to write this in fudje is basically to manually pull the `(setup ...a... ...b... ...c...)` expression out of the fact and replace the meta-constants with keywords (just prepend `:`). Moreover, since now the call to `setup` is outside the scope of  the `fact, there are 2 potential problems. First of all any mocking we do doesnt' affect the call to`setup`, and`setup`might not actually work with keywords! How do we fix this? With a`delay`...here is the fudje version:

``` clj
(let [s (delay (setup :...a... :...b... :...c...))]
(fact "testing whatever" 
  (do @s (f1 ...a...)) => :some-value
  (f2 ...b...) => :some-other-value  
))
```

With the above code, we can even mock parts of `setup` because the actual expression is evaluated _within_ the scope of a `fact`. Admittedly, this is slightly more involving and not as intuitive as midje, but in my experience this is very rarely needed, and so the tradeoff is rather small. 

Now, you need to do this dance if, and only if, your `setup` expression involves meta-constants AND needs to be 'mockable' (needs to be able to be affected by the mocks). 

If your `setup` expression does none of that, you can simply pull it outside. This doesn't break midje, while it allows fudje to work as expected. So you could re-write the above expression you pasted as:

``` clj
(do (setup obj)
(fact 
   (call obj) => <check>))
```

I personally find this much better...Why does the setup need to be part of the assertion? The real answer is "because midje allows it" right? Well, fudje takes a different view. A test should be concerned only with assertions and potentially mocks - nothing else...

And actually , if as you say you do this a lot, you could write a super simple macro `fact-with-setup` which does this simple transformation. I'm not too convinced that this belongs in fudje though. I'll have think about that...maybe some meta-data on the `fact` symbol could work as well...I'll keep you posted :)

On your second comment about isolated forms, why would you want to do this? What value is there in a fact which doesn't have any assertions?Am i missing something?

**Comment by hoebat** [comment]
It's more for selfish reasons... I have written a document generation from tests and it's quite a bit of work to change all the test code.

For example:
http://docs.caudate.me/hara/hara-event.html#normal

is generated via:
https://github.com/zcaudate/hara/blob/master/test/documentation/hara_event/strategies.clj#L11-L19

and... it is possible to go in and change everything but it's gunna take a long time. I'd prefer if it was supported so that it is just compatible

I've also had to change how the `let` forms work:
https://github.com/zcaudate/vinyasa/blob/master/test/vinyasa/reflection_test.clj#L27-L35

but because there is manipulation of state within a single fact form, it causes problems. 
I would prefer to have an implicit `do` when two or more statements are seen as opposed to have to be explicit about it. 

I can take a look at it if you like. There is an example of how I used a similar grouping method to parse for code blocks in `facts` forms:
https://github.com/helpshift/hydrox/blob/master/src/hydrox/doc/parse.clj#L80

Basically, it's about recognising the arrow forms like `=>`, `=not=>` and `=throws=>` and the grouping the forms around them, sorting them and adding a `do` form around the input that has multiple forms

**Comment by jimpil** [comment]
You've gone through a lot of trouble to make this work, and things are quite tied up to midje. I think the amount of code you've had to write + the amount of code Brian (midje author) had to write stems directly from the fact that midje tries to be way too flexible. I really don't want to have to write a parser and stuff in order to support what midje supports. In my view, the flexibility midje offers can often be a nuisance both to the implementor and its users. I mean look at all the hoops you had to jump through. It seems to be that if you had stuck with clojure.test based tests, you could have used meta-data on regular Vars and/or functions to end up with a simpler implementation. Please don't take this the wrong way - i'm trying to be helpful, but the thing is that what you want to support strays away from the core principles behind fudje (simplicity & focus). What i could do perhaps is to provide a polymorphic pre-processing hook to `fact` via meta-data. So for example, i'd define a `preprocess` multi-method with a :default pass-through implemantation, and you would be able to write something like this:

``` clj
(^:with-setup fact "some-description" 
  (setup obj) 
  (call obj) => <check>
) 
```

But it would be up to you to provide the implementation for `preprocess` for the `:with-setup` case. Your implementation would basically take the 2nd arg and pull it outside in a `do` or a `let` and then delegate to regular fact (no meta-data). How does that sound? Would that  help at all?

**Comment by hoebat** [comment]
that’s not going to help unfortunately…. =/

it just means I have to go through every single one of those forms and add the ^:with-setup meta… which is worse than before.

**Comment by jimpil** [comment]
hmm I see... The intro.md does mention that some manual rewrite will probably be required ,especially if you're are a midje power-user. I'm really sorry but I'm not sure what else I can do at this point. You see, your tests are sort of free-form (intermingled with some sort of mark-up descriptions), and that is a problem for fudje. The fact that midje allows for such things is bizarre, if you ask me, and therefore I have to admit I am very reluctant doing the same for fudje.  Feel free to try it out in a fork of yours, if you're so inclined, but I suspect you'll find things will start getting messy pretty quickly. Apologies for potentially disappointing you, and all the best of luck in however you decide to proceed with this.

**Comment by hoebat** [comment]
I've experimented here:

https://github.com/zcaudate/lucid/blob/master/src/lucid/sweet.clj#L172-L193
